# rates_m6A

**"Slow RNAPII transcription elongation rate, low levels of RNAPII pausing, and elevated histone H1 content at promoters associate with higher m6A deposition on nascent mRNAs"**  

Alicia Gallego, José Miguel Fernández-Justel, Sara
Martín-Vírgala, Magdalena M Maslon, María Gómez

contact: agallego@cbm.csic.es, mgomez@cbm.csic.es  

## SEQUENCING EXPERIMENTS INCLUDED:  
- **4sUDRBseq**: scripts for processing transcription elongation rates (data from this work, some scripts based on Fuchs et al. 2014).  
- **ChMeRIPseq**: scripts for processing information about m6A levels in chromatin-enriched RNAs (ChMeRIP) (data from Fernandez-Justel et al. 2021).  
- **CheRNAseq**: scripts for processing expression levels in chromatin-enriched RNAs (cheRNA) (data from Fernandez-Justel et al. 2021).
- **Elongation_rate_data_Maslon**: scripts for processing information about transcription elongation rates (data from Maslon et al. 2019) and m6A levels across Ensembl protein-coding genes.  
- **H1c_H1d_data_Cao**: scripts for processing information about levels of H1c and H1d histone variants (data from Cao et al. 2013) across RefSeq protein-coding genes. 
- **Histone_marks_data_Geeven**: scripts for processing information about histone marks H3K4me3, H3K9me3, and H3K27me3 (data from Geeven et al. 2015) across RefSeq protein-coding genes. 
- **METTL3_data_Xu**: scripts for processing information about METTL3 levels (data from Xu et al. 2021) across RefSeq protein-coding genes. 
- **MNase_seq**: scripts for processing positioned nucleosomes (data from this work) across RefSeq protein-coding genes.    
- **RNAPII_ChIPseq**: scripts for processing RNAPII occupancies and RNAPII pausing indexes (RNAPII-ChIP sequencing) (data from Fernandez-Justel et al. 2021).  


## REFERENCES   
· Cao K., et al., 2013. High-Resolution Mapping of H1 Linker Histone Variants in Embryonic Stem Cells. PLOS Genet. 9, e1003417. https://doi.org/10.1371/journal.pgen.1003417  
· Fernández-Justel JM., et al., 2021. Histone H1 regulates non-coding RNA turnover on chromatin in a m6A-dependent manner. https://doi.org/10.1101/2021.10.12.464039  
· Fuchs G., Voichek Y., et al., 2015. Simultaneous measurement of genome-wide transcription elongation speeds and rates of RNA polymerase II transition into active elongation with 4sUDRB-seq. Nat. Protoc. 10, 605–618. https://doi.org/10.1038/nprot.2015.035  
. Geeven G., et al. 2015. Local compartment changes and regulatory landscape alterations in histone H1-depleted cells. Genome Biol 16, 289. https://doi.org/10.1186/s13059-015-0857-0  
· Maslon MM., et al., 2019. A slow transcription rate causes embryonic lethality and perturbs kinetic coupling of neuronal genes. EMBO J. 38, e101244. https://doi.org/10.15252/embj.2018101244
· Xu W., et al., 2021. METTL3 regulates heterochromatin in mouse embryonic stem cells. Nature 591, 317–321. https://doi.org/10.1038/s41586-021-03210-1  
